sway-lib-std/src/vec.sw

```diff
@@ -208,4 +208,22 @@ impl<T> Vec<T> {
         self.len -= 1;
         Option::Some(read(self.buf.ptr() + self.len * size_of::<T>()))
     }
+
+    /// Swaps two elements
+    /// a - The index of the first element
+    /// b - The index of the second element
+    /// Panics if `a >= self.len || b >= self.len` .
+    pub fn swap(mut self, a: u64, b: u64) {
+        assert(a < self.len);
+        assert(b < self.len);
+
+        let val_size = size_of::<T>();
+
+        let a_ptr = self.buf.ptr() + a * val_size;
+        let b_ptr = self.buf.ptr() + b * val_size;
+
+        let a_val = read(a_ptr);
+        copy(b_ptr, a_ptr, val_size);
```

Comment:
I suspect this call to `copy` will panic if `a` and `b` are the same. We should probably just return early if `a == b` and not do anything.

Reply:
I don't see why it would panic however I agree that if `a == b` then that is a noop. Then again how often is someone going to swap the same elements? Is it worth adding an additional check and expanding the bytecode a little bit more just to do a noop that otherwise is a mistake from the user (that I do not immediately see from looking at the implementation of copy why it would panic)? In terms of efficiency and bytecode size I would argue that there is no reason to include it unless it actually panics and if it panics then perhaps that is a compiler issue that needs to be solved instead of a high level library bandaid.

Reply:
The reason I expect a panic is because `copy` calls `mcp` under the hood which does not allow overlapping ranges (ref: https://github.com/FuelLabs/fuel-specs/blob/master/specs/vm/instruction_set.md#mcp-memory-copy). I do agree that an additional check will cause extra bytecode that is not going to be needed in most cases, but we have to handle that case in a public API unfortunately. If we don't want to do this here, then we have to change the spec (and implementation) of `mcp`.

Reply:
Ref: https://github.com/FuelLabs/fuel-vm/issues/147